Improve target CNAME resolving for cloaking

lifenjoiner opened the pull request:
Avoid bypass and increase privacy.

Issues with CNAME cloaking rules:
* Rule bypass if upstream server fails
* Privacy leak if `net.LookupIP`/`net.LookupHost` flows to plain DNS
* Out of service if clients query AAAA records but IPv6 connection is not available

lifenjoiner committed:
Avoid cloaking rule bypass

CNAME does not exist, or upstreaming errors occur.

lifenjoiner committed:
Improve target CNAME resolving for cloaking

* Avoid using plain DNS
Use only encrypted servers. These are users' queries.
scenarios:
Server host uses plain DNS as the system DNS

* Limits (on Windows) of `net.LookupIP` and `net.LookupHost`:
It only queries IPv4 via IPv4 and IPv6 via IPv6, not all servers have
dual stack network interface.

* Wrap new functions `resolve` and `resolveEncrypted` that return IP array

* Align `resolve*` function interfaces

jedisct1 commented:
What does that do exactly?

lifenjoiner commented:
* Avoid cloaking rule bypass
https://github.com/DNSCrypt/dnscrypt-proxy/blob/e9d11c5d54c546f61a281ca83c271ae862b7a637/dnscrypt-proxy/example-cloaking-rules.txt#L12
If there is an unexpected error from `net.LookupIP`, the rule will be bypassed, silently.

* Improve target CNAME resolving for cloaking
https://github.com/DNSCrypt/dnscrypt-proxy/blob/e9d11c5d54c546f61a281ca83c271ae862b7a637/dnscrypt-proxy/example-dnscrypt-proxy.toml#L444
When doing CNAME flattening. `net.LookupIP` uses system DNS. If dnscrypt-proxy is serving on a server (router?) which does not use dnscrypt-proxy as the system DNS, doesn't it use the plain DNS?

jedisct1 commented:
To be honest, I'm super scared about touching anything in `xtransport`, as any functional changes there can have too many implications.

lifenjoiner commented:
We test it and improve it.

This only reuses the host name resolving part for DoH stamps.

It works for both:
```
[s] [NOTICE] dnscrypt-proxy service is not usable yet
[s] [NOTICE] Resolving server host [dns.nextdns.io] using bootstrap resolvers over udp
[s] [NOTICE] [dnscry.pt-lasvegas-ipv4] OK (DNSCrypt) - rtt: 138ms
[s] [NOTICE] [dnscry.pt-tokyo-ipv4] OK (DNSCrypt) - rtt: 232ms
[s] [INFO] [nextdns-ultralow] TLS version: 304 - Protocol: h2 - Cipher suite: 4865
[s] [NOTICE] [nextdns-ultralow] OK (DoH) - rtt: 262ms
[s] [NOTICE] Sorted latencies:
[s] [NOTICE] -   138ms dnscry.pt-lasvegas-ipv4
[s] [NOTICE] -   232ms dnscry.pt-tokyo-ipv4
[s] [NOTICE] -   262ms nextdns-ultralow
[s] [NOTICE] Server with the lowest initial latency: dnscry.pt-lasvegas-ipv4 (rtt: 138ms)
[s] [NOTICE] dnscrypt-proxy is ready - live servers: 3
[s]   127.0.0.1       forcesafesearch.google.com      A       PASS    157ms   dnscry.pt-lasvegas-ipv4
[s]   127.0.0.1       www.google.com  A       CLOAK   161ms   -
[s] [NOTICE] Stopped.
```

jedisct1 commented:
There are too many changes to xTransport here.

lifenjoiner commented:
Made some mistakes. I'm amending it.

lifenjoiner committed:
Improve cache for cloaking CNAME

lifenjoiner committed:
Refactor resolveUsingResolver

lifenjoiner committed:
Set longer resolving ReadTimeout for resolveUsingResolver

Reduce retries. Default 2s is short.
Upstream CNAME resolving may be slow (3+ tiers, first seen?):
dig -t aaaa www.bing.com

lifenjoiner committed:
Append systemd assigned listening address for internal queries

lifenjoiner committed:
Extend PluginCloak read locking scope

lifenjoiner commented:
I added descriptions to commits. And here is a summary of key points:
* CNAME cloaking rule bypass: parental control bypass
* `net.LookupIP`/`net.LookupHost` may use plain DNS: privacy leak
* `net.LookupIP`/`net.LookupHost` only queries IPv4 via IPv4 and IPv6 via IPv6: won't get IPv6 if ...
* Scenarios: Cross-device services
* Query and cache IPv4 and IPv6 separately
* Read locking covers reading the cache
* Upstream multi-tier CNAME resolving needs more time
* Update `internalResolvers` for systemd socket mode
* Align `resolve*` function interfaces, and adapt them for reuse
* Compact `resolveUsingResolver`

lifenjoiner committed:
Simplify resolveEncrypted

lifenjoiner commented:
Issues with CNAME cloaking rules:
* Rule bypass if upstream server fails
* Privacy leak if `net.LookupIP`/`net.LookupHost` flows to plain DNS
* Out of service if clients query AAAA records but IPv6 connection is not available

Reasoning of the patches:
Privacy ==> Use `internalResolvers` ==> Reuse `resolveUsingResolvers()` ==> Return `[]IP` ==> Adapt and unify `resolve*` functions.
Logic and results for `resolveAndUpdateCache` keep working.

Does it have any regressions?

jedisct1 commented on dnscrypt-proxy/systemd_linux.go:
```diff
@@ -9,6 +9,15 @@ import (
+	}
+	return false
+}
+
```
You can use `slices.Contains` for this.

lifenjoiner replied:
Thank you!
I thought using a self written function in this less tasks case is better than importing a new package.

lifenjoiner committed:
Use `slices.Contains`

lifenjoiner committed:
A minor fix for `resolveUsingSystem`

IPv4.To6() is not nil.
Make it work as expected with only `ipv6_servers`.
IPv4-mapped IPv6 addresses work only in dual-stack networks, not pure IPv6.

jedisct1 commented:
Thanks!
